plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -234,6 +256,129 @@ public void onFailure(Exception e) {
+                ingestPipelineResponse.pipelines(),
+                modelId,
+                org.opensearch.ingest.PipelineConfiguration::getConfigAsMap,
+                org.opensearch.ingest.PipelineConfiguration::getId
+            );
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            allDependentPipelineIds.size()
+                                + " ingest pipelines are still using this model, please delete or update the pipelines first: "
+                                + Arrays.toString(allDependentPipelineIds.toArray(new String[0])),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkSearchPipelineBeforeDeleteModel(String modelId, ActionListener<Boolean> actionListener) {
+        GetSearchPipelineRequest getSearchPipelineRequest = new GetSearchPipelineRequest();
+        client.execute(GetSearchPipelineAction.INSTANCE, getSearchPipelineRequest, ActionListener.wrap(searchPipelineResponse -> {
+            List<String> allDependentPipelineIds = findDependentPipelines(
+                searchPipelineResponse.pipelines(),
+                modelId,
+                org.opensearch.search.pipeline.PipelineConfiguration::getConfigAsMap,
+                org.opensearch.search.pipeline.PipelineConfiguration::getId
+            );
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            allDependentPipelineIds.size()
+                                + " search pipelines are still using this model, please delete or update the pipelines first: "
+                                + Arrays.toString(allDependentPipelineIds.toArray(new String[0])),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkDownstreamTaskBeforeDeleteModel(String modelId, Boolean isHidden, ActionListener<DeleteResponse> actionListener) {
+        CountDownLatch countDownLatch = new CountDownLatch(3);
+        AtomicBoolean noneBlocked = new AtomicBoolean(true);
+        List<String> errorMessages = new ArrayList<>();
+        ActionListener<Boolean> countDownActionListener = ActionListener.wrap(b -> {
```

Comment:
Not quite following, what's the case that can't adapt to? Do you have an example?

Reply:
Hey @zane-neo dont have an example of the top of my head but this looks like it will probably be reused for a different resource later. Currently it does searching for agent, search and ingest pipeline; my thought process is that it could possibly do this for a different resource (when a new process is involved that uses model_id). Or if a different resource we have has a dependency like a connector id we could use this method similarly just a suggestion to think about.

Reply:
Currently I prefer to keep it in this way since we don't have another example and we're not really sure about the generic scenario. If in the future we have another one, we could design a method for that.